docs/modules/SDDs/pages/0021-cluster-catalog-compilation.adoc

```diff
@@ -0,0 +1,170 @@
+= SDD 0021 - Cluster Catalog Compilation
+
+:sdd_author:    Simon Rüegg
+:sdd_owner:
+:sdd_reviewers:
+:sdd_date:      2020-06-10
+:sdd_status:    draft
+include::partial$meta-info-table.adoc[]
+
+[NOTE]
+.Summary
+====
+This document describes how, where and when the xref:0006-commodore-configuration-catalog-compilation.adoc[catalog compilation engine] runs.
+====
+
+
+== Motivation
+
+The compilation of a cluster catalog is a central piece of the configuration management within project Syn.
+It generates the actual configuration for a cluster based on various inputs and configurations.
+It's therefore crucial to have a robust and scalable system in place to run these compilation jobs in a secure and reactive manner.
+
+
+=== Goals
+
+* Define where (in terms of which cluster, namespace, etc.) the jobs are run
+* Define what should cause the recompilation of the catalog for a certain cluster
+* Scale to several hundrets and with future changes to thousands of clusters
+
+
+=== Non-Goals
+
+* The handling of a compiled catalog (for example merge request automation) is out of scope for this design document
+* Solution to automate the component updates
+
+
+== Design Proposal
+
+The design splits the problem into two parts:
+
+. Figuring out which cluster catalogs need to be compiled
+. Running the actual compilation job
+
+
+=== Quartermaster
+
+A new tool called https://en.wikipedia.org/wiki/Quartermaster#Royal_Navy[Quartermaster] is implemented.
+This tool should solve the problem of figuring out which cluster catalogs need to be recompiled.
+The tool gets input in the form of webhooks which are triggered by Git repositories or other tools.
+By using the Kapitan inventory the tool needs to figure out for which clusters the config changed and therefore needs a recompilation.
+In a first step this could be relatively rudimentary and only use the input information from a webhook (cluster/tenant ID) to decide which clusters to compile.
+
+The information if a cluster needs to be compiled is then stored in the cluster object of the Lieutenant API.
+In a first PoC phase this can be an annotation like `cluster.syn.tools/needs-compilation=true`.
```

Comment:
Maybe we should store a timestamp here. By doing so, the executing part does not need to update the object.

* Quartermaster figures a rebuild is needed and writes the current timestamp into the annotation.
* The thing responsible for trigger the build jobs, compares the timestamp in the annotation with the timestamp of the last execution. If the last execution is older than what is stored in the annotation, a new build gets triggered.

Reply:
But then we don't know if the job ran or not.
Also we just move the storing of information from the API to the clusters, which probably makes it harder to do.

Reply:
Is it really the responsibility of Quartermaster to know which job ran and which not? This is something monitoring is for. Besides that, how would that true/false flag help Quatermaster to do its decision? What information are you talking about that would need to be stored on the cluster?

Reply:
Outcome of the discussion:
We need to know for which "compile request" the compilation was run/needs to run. So neither a boolean nor a timestamp are the right data types I guess.

We could use a unique ID (i.e. compile-ID) which changes for each compile request. That way the Quartermaster can just bump the ID if a new compile needs to run and the system that runs the jobs knows if for a certain request the compile was already run.

Reply:
Works for me. Just a side note: in this context, a unique ID and a timestamp are equivalent.

Reply:
Almost: a unique ID can be made unique, while a timestamp might collide (if created at the same time).

Reply:
Please update the text. Right now, it still specifies that we use a boolean and we agreed to not do so.